feat(coral): Add Request a new dropdown navigation in Header

mathieu-anderson opened the pull request:
## About this change - What it does

- Add `Request new` dropdown in Header to directly link to the form for creating:
  - Topic : redirect to existing route `/topics/request`
  - ACL: redirect to *new* route `/request/acl`
  - Schema: redirect to *new* route `/request/schema`
  - Connectors: redirects to existing route `connectors/request`
- To allow for rendering the ACL and Schema request forms without having access to a topic in the URL params, add:
  - ACL: `isSubscription` prop on `TopicConsumerForm` and `TopicProducerForm` is set to `true` when we have access to a topic name in URL params -> the topic name field is disabled (cannot switch topics in this context). 
  - Schema: if `topicName` picked from params is undefined, render enabled topic names Select
 
## New topic request

### New Header link

https://user-images.githubusercontent.com/20607294/236400399-6e79cea6-1e4b-4e1d-8a1f-eeeadb42b835.mov

### UI navigation

https://user-images.githubusercontent.com/20607294/236400839-df1f071b-7ad9-4fa1-89d0-00d1fb05ff56.mov



## New ACL request
  
### New Header link

https://user-images.githubusercontent.com/20607294/236400433-496294fc-9a75-4b22-b01e-273d4c84bfe9.mov

### UI navigation (should be from Topic overview which is not implemented)

https://user-images.githubusercontent.com/20607294/236401896-2aadaaef-90fa-4ba8-b9d5-1467fcc70d14.mov


## New Schema request

### New Header link


https://user-images.githubusercontent.com/20607294/236400491-2e6f39bb-c6df-4747-80f8-037a97326708.mov

### UI navigation (should be from Topic overview which is not implemented)


https://user-images.githubusercontent.com/20607294/236402187-6c7ff99f-57f0-4879-88c2-616309b9b47e.mov



## New Connector request

### New Header link

https://user-images.githubusercontent.com/20607294/236400543-287efc35-15ef-49e2-8b45-76fae8b24a0c.mov

### UI navigation

https://user-images.githubusercontent.com/20607294/236402304-63186803-bd1d-48b1-b386-fd2c68e8385a.mov



Resolves: #1114

programmiri commented:
General not about navigating / using with keyboard and screenreader: 

- I think keyboard navigation works fine! :D 
- Screenreader usage is a bit suboptimal. The dropdown is read out correctly, which is already good! 
but SR users don't now that this is part of the navigation and it does go to a page with a form. With a link in a navigation, it's pretty clear what is happening, but not this way. I guess we can't change the elements being used (making it a proper `nav` with `a href` would already solve this mostly)? (At least add an extra information with a `visually-hidden` span or aria-label, so SR users know this not not (only) a "Request new" collapsed menu, but actually acts as navigation to these different forms, could help a bit hopefully)

programmiri commented:
Hm I think we need to check out a few different cases. I'm logged in with team DevRel and have different (or none_) topics / env then you currently, and the ACL form does not work very smooth for me. 

I can choose an environment where it seems I have no topics and the form just does not change, which feels like I could fill it out, which I can't, because there's not topic :D 

And in this recording, after I've chosen the dev env, the other option for env was just gone and I could not reset that in any way. 

https://user-images.githubusercontent.com/943800/236470786-c5c7e550-3324-46dd-b151-ef23a57859d3.mov



Schema form looks ok, but I have schemas in the one env that I can chose. But as far as I can see, if I don't have a topic in that env where I can register a schema, I would just see the skeleton loading without more information?

mathieu-anderson commented:
> I can choose an environment where it seems I have no topics and the form just does not change, which feels like I could fill it out, which I can't, because there's not topic :D

That is pretty weird, but I have seen that state before, so it should be a bug :o

> And in this recording, after I've chosen the dev env, the other option for env was just gone and I could not reset that in any way.

That is expected, it's because the topic you chose only exists in this one env. So you can't choose a different env after choosing this topic ^^ If you choose a topic that is in different envs, you will be able to switch.

> but I have schemas in the one env that I can chose. But as far as I can see, if I don't have a topic in that env where I can register a schema, I would just see the skeleton loading without more information?

Not sure what you mean :o

mathieu-anderson commented:
I'm thinking smartness that makes it so that:
- only environments which have the currently selected topic are in the options
- only the topics that are in the currently selected environments are in the option

Is responsible for the weirdness in your video

programmiri commented:
> That is expected, it's because the topic you chose only exists in this one env. So you can't choose a different env after choosing this topic ^^ If you choose a topic that is in different envs, you will be able to switch.

I thought so, but I'm not sure if it's ok to leave it like that, because I'm not sure users will understand 🤔  

> Not sure what you mean :o

(regarding schemas): In `TopicSchemaRequests`, when the `topicNames` is `undefined`, it's showing the loading native select. If something goes wrong with `getTopicNames` (list will be undefined then and we've no error handling) we will still show the loading animation. And if the list is empty, there's an empty select element without more info. 

I think we need designs (or decisions on designs) for different scenarios 🤔  here.

mathieu-anderson commented:
> (regarding schemas): In TopicSchemaRequests, when the topicNames is undefined, it's showing the loading native select. If something goes wrong with getTopicNames (list will be undefined then and we've no error handling) we will still show the loading animation. And if the list is empty, there's an empty select element without more info.

Ah I see! Tbh I didn't think that the topic list could ever be empty. Would that be a frequent occurence? But I do see the issue if something goes wrong with the call to `getTopicNames` ^^

programmiri commented:
> I mean, presumably a user knows in which envs the topic is, so I don't think they would be surprised :o

In this cases, I think it's better to not rely on users knowing / remembering things and (where possible) don't make them think at all. I would say we should clarify that with UX.  Also, in that caae the form doesn't do anything, so at least an information for users (aka `No topics in this environment, please choose a different one` or something like that may be helpful. 

> Tbh I didn't think that the topic list could ever be empty. Would that be a frequent occurence? 

I don't know, so we would need to clarify if that's a case that could happen. And even if that shouldn't happen, there could be an error from the Api for some reason & that should be covered.

mathieu-anderson commented:
> Also, in that caae the form doesn't do anything, so at least an information for users (aka No topics in this environment, please choose a different one or something like that may be helpful.

Yes for sure ^^ It is not harder to do this than to remove them from the list.

> And even if that shouldn't happen, there could be an error from the Api for some reason & that should be covered.

Also for sure ^^ Not asking about frequency to use it ass an excuse to not to do it, just to have a clearer idea of the possible app states ^^

mathieu-anderson commented:
@programmiri :
- added disabled options instead of removing the options for unavailable envs: [b2be104](https://github.com/aiven/klaw/pull/1119/commits/b2be1040ec0fb4676e863fcf2f47b982eae14d7c)
- added a more explcicit UI regarding the Topic names / Prefix field [adcb000](https://github.com/aiven/klaw/pull/1119/commits/adcb000520a37d5a36c037030f2ab8bf7976dd87)
 

https://user-images.githubusercontent.com/20607294/236823962-814dfb7b-cd1b-4776-a312-3d3b8d7e6def.mov

- added `Alert` components rendered when there is an unforeseen issue while fetching topic names / environments in Schema request form [303d347](https://github.com/aiven/klaw/pull/1119/commits/303d347509b2eee7fc2b3cc96ce3c80f8fd2ed90)

<img width="1491" alt="Screenshot 2023-05-08 at 14 25 00" src="https://user-images.githubusercontent.com/20607294/236824235-6f608bab-b736-4989-94e3-4e12014f08dd.png">
<img width="1500" alt="Screenshot 2023-05-08 at 14 26 22" src="https://user-images.githubusercontent.com/20607294/236824240-39eb7d3d-1692-4a99-ac22-9a59e2c5a7fa.png">

mathieu-anderson commented:
> When I want to subscribe as a producer, I get that information that I've to choose env etc. first. But when I want to subscribe as a consumer, I don't get that -> I think that should be consistent.

The reason why this is needed in Producer form and not in Consumer form is that we need to know if an environment is an Aiven cluster or not to know what to render for the pattern type and topic name fields:
- if `isAivenCluster`: pattern type is disabled (can only be`LITERAL`)
- is `!isAivenCluster`: pattern type is enabled, and user needs to choose either `LITERAL` or `PREFIXED`. which then renders:
  - if `LITERAL`: select component with topics in the environment as options
  - if `PREFEIXED`: text input component
  
All of this does not apply to the Consumer form, which can be simpler.

> When I want to subscribe as producer and by mistake chose the wrong environment, I can't change that (I think I could if I have chosen a topic that belongs to other environments too, but then I would only be able to chose these envs and not all I can theoretically choose?). Maybe a we need a way to reset this required filters?

I think this is a UX choice. We can either have the current very strict state, or we can allow to switch env, and reset the topic name field if the topic does not exist in the new environment. Which can also conceivably be very confusing.

>  Environment has no topics suddenly?

Facepalm,gif. That is 100% my mistake (the "placehodler" field was still registered with the form, and sending `"Select environment and topic pattern type first"` as value for topic name). Fix here: [b7c7541](https://github.com/aiven/klaw/pull/1119/commits/b7c75418961e462e42186eca09687694e4c0416d)


> It seems I can't subscribe anywhere as a consumer (logged in in team devrel) but it's not made clear. The envs are all disabled, but two of them with "unavailable for selected topic" even so there's no topic selected.

That is the weirdest of them, I would like more info about it ^^

programmiri commented:
>I think this is a UX choice. We can either have the current very strict state, or we can allow to switch env, and reset the topic name field if the topic does not exist in the new environment. Which can also conceivably be very confusing. Or a third option! Maybe fusing the environment and topic name fields, and using a ComplexNativeSelect or Combobox to render options like Topicname [DEV], Topicname [TST] ? ... that would need more UI research

Hm that's tricky 🤔  being strict is good, but also gets me stuck as a user :D which I don't like (am a very impatient user). Probably best to sync with Mustafa on that and see how different scenarios are playing out vs. how they should work. 


>That is the weirdest of them, I would like more info about it ^^ 
I'll make a recording after the meetings!  I thought I did, but it seems I forgot uploading it and (of course) already cleared my desktop 🙈

mathieu-anderson commented:
>  being strict is good, but also gets me stuck as a user 

I mean, as I imagine it, the flow would be not really a "stuck" flow. If you want to change env after picking a topic, and you see that there is only one env available, I imagine a user would:
- realize this topic only exist in this env and be like "ah well, guess I need to do it in this env"
- realize they picked the wrong topic and pick the one that actually exists in the env they want to switch to

No?

programmiri commented:
>I mean, as I imagine it, the flow would be not really a "stuck" flow. If you want to change env after picking a topic, and you see that there is only one env available, I imagine a user would:

- realize this topic only exist in this env and be like "ah well, guess I need to do it in this env / I need to promote the topic to a different env"
- realize they picked the wrong topic and pick the one that actually exists in the env they want to switch to
No?


I think I have to test this with the updated env endpoints, because for me at the moment I do get stuck, because after choosing a environment for DEV and getting a list of topics for DEV, the option TST in env get's disabled (bc the first topic is preselected in the list, which does not exist on TST).  (this is the workflow for a user choosing the wrong env by mistake now :D)

Now I only can get TST enabled again if I find a topic that exists in TST, too 🤔

programmiri commented:
> > It seems I can't subscribe anywhere as a consumer (logged in in team devrel) but it's not made clear. The envs are all disabled, but two of them with "unavailable for selected topic" even so there's no topic selected.

> That is the weirdest of them, I would like more info about it ^^


OK I got it. Tried it signed in for STAGINGTEAM

https://github.com/aiven/klaw/assets/943800/46d001e3-b126-4e70-bf51-85a1d83e1d09

mathieu-anderson commented:
>  (this is the workflow for a user choosing the wrong env by mistake now :D)

I would imagine that they know which Topic they want to create an ACVL request for. Therefore, if they mistakenly choose `DEV` instead of `TST` (or any "higher level" env in the promotion system, which goes like `DEv` -> `TST` -> `ACC` -> `PROD`, or similar), the topic they want will *still* be available in the `DEV` environment. They can choose it, and then change it to `TST`.

> Now I only can get TST enabled again if I find a topic that exists in TST, too

Yes, that it how it is supposed to work ^^

mathieu-anderson commented:
<img width="876" alt="Screenshot 2023-05-10 at 08 44 18" src="https://github.com/aiven/klaw/assets/20607294/fd17a585-8a97-40bf-8524-3f388bda2ab4">

Ah yes! I saw this bug before and forgot to address it. I imagine in this instance the `selectedTopic` is `undefined`, which does not exist in any env haha

mathieu-anderson commented:
@programmiri Whew that was a rabbit hole. I implemented the more consistent UI with placeholder until env is chosen here: [4c24c9e](https://github.com/aiven/klaw/pull/1119/commits/4c24c9eb3819ab004d6b11ee84617b6e62b03ef7)

But more importantly, I replaced the behaviour of `NativeSelect` setting `""` as the field value on blur by a trigger to validate: [e9a51cf](https://github.com/aiven/klaw/pull/1119/commits/e9a51cf67b2af11c28664f81246f1ee587d50c23)

- we do not set a empty value in the form, which was the root cause for the issue (`""` makes no sense in most contexts for a required field):
https://github.com/aiven/klaw/assets/20607294/906cb999-6a0f-42d1-9ba8-7f29f93a3f6b

- we use `errorMap` instead of validating string length to still render an error when validating required `NativeSelect`

Seems like the issue is now fixed:
https://github.com/aiven/klaw/assets/20607294/67f88027-878b-4583-bf8b-d858d5ffe09c

But this illustrates how having env and topic name as separate fields might be a bad idea when:
- they so strongly depend on one another
- make no sense individually

Will create a "future improvement" issue to reflect on this.

programmiri reviewed (commented):
I left a few comments. Let's pack this behind a feature flag 🥷  and then we can merge it. 

Though it has to be rebased.

mathieu-anderson commented on coral/src/app/components/Form.tsx:
```diff
@@ -408,7 +408,19 @@ function _NativeSelect<T extends FieldValues>({
             disabled={disabled || isSubmitting}
+            aria-readonly={props.readOnly}
             onBlur={(option) => {
+              // Prevent from setting empty string value in form when blurring out of the component
```
For a not required select, we do not register any value in the form ^^ Essentially, this means that a form would have `undefined` as the value for the non required field, because the form object won't have a property for the optional field until one is chosen.

programmiri replied:
💡  true!

programmiri commented on coral/src/app/layout/header/HeaderNavigation.tsx:
```diff
@@ -1,37 +1,84 @@
+            <HeaderMenuLink
+              icon={notifications}
+              linkText={"Go to approve requests"}
+              href={`/approvals/topics`}
```
Oh gosh I didn't even notice that part 😂  yes it should! Another nice catch! :D

programmiri commented on coral/src/app/layout/header/HeaderNavigation.test.tsx:
```diff
@@ -1,13 +1,20 @@
 
+const isFeatureFlagActiveMock = jest.fn();
+
+jest.mock("src/services/feature-flags/utils", () => ({
```
I think it could be a good idea to actually add the feature flag that we need to be active 🤔  Not because it's important for the test, but when we remove the flag, it will come up when searching for `FEATURE_FLAG_TOPNAV_DROPDOWN` and we'll know that we can safely remove the mock (or part of it, in case there are multipe flags mocked) 

Not important for this PR now, but if you agree, I would later do a small follow up on the docs to describe it.

mathieu-anderson replied:
Yes that makes sense :o
